### PR TITLE
Added classic CQRS flow in .NET

### PR DESCRIPTION
**Pull request by oskardudycz**
Added E-Commerce sample showing the classic Event Sourcing and CQRS flow with EvenStoreDB in .NET.

## Overview
- typical Event Sourcing with CQRS flow.
- stores events to EventStoreDB.
- shows how to organise the write model and read model handling.
- Builds read models using [Subscription to `$all`](https://developers.eventstore.com/clients/grpc/subscribing-to-streams/#subscribing-to-all).
- Read models are stored as ElasticSearch documents.
- Shows how to unit and integration test solution.

## Write Model
- Provides the basic boilerplate together with Core projects,
- [EventStoreDBRepository](./Core/Core.EventStoreDB/Repository/EventStoreDBRepository.cs) repository to load and store aggregate state,
- [IProjection](./Core/Core/Projections/IProjection.cs) interface to handle the same way stream aggregation and materialised projections,
- Thanks to that added dedicated [AggregateStream](./Core/Core.EventStoreDB/Events/AggregateStreamExtensions.cs#L12) method for stream aggregation
- See [sample Aggregate](./Carts/Carts/Carts/Cart.cs) and [base class](./Core/Core/Aggregates/Aggregate.cs)

## Read Model
- Read models are rebuilt with eventual consistency using subscribe to all EventStoreDB feature,
- Added hosted service [SubscribeToAllBackgroundWorker](./Core/Core.EventStoreDB/Subscriptions/SubscribeToAllBackgroundWorker.cs) to handle subscribing to all. It handles checkpointing and simple retries if the connection was dropped.
- Added [ISubscriptionCheckpointRepository](./Core/Core.EventStoreDB/Subscriptions/ISubscriptionCheckpointRepository.cs) for handling Subscription checkpointing.
- Added checkpointing to EventStoreDB stream with [EventStoreDBSubscriptionCheckpointRepository](./Core/Core.EventStoreDB/Subscriptions/EventStoreDBSubscriptionCheckpointRepository.cs) and dummy in-memory checkpointer [InMemorySubscriptionCheckpointRepository](./Core/Core.EventStoreDB/Subscriptions/InMemorySubscriptionCheckpointRepository.cs),
- Added [ElasticSearchProjection](./Core/Core.ElasticSearch/Projections/ElasticSearchProjection.cs) as a sample how to project with [`left-fold`](https://en.wikipedia.org/wiki/Fold_(higher-order_function)) into external storage. Another (e.g. ElasticSearch, EntityFramework) can be implemented the same way.

## Tests
- Added sample of unit testing in [`Carts.Tests`](./Carts/Carts.Tests):
  - [Aggregate unit tests](./Carts/Carts.Tests/Carts/InitializingCart/InitializeCartTests.cs)
  - [Command handler unit tests](./Carts/Carts.Tests/Carts/InitializingCart/InitializeCartCommandHandlerTests.cs)
- Added sample of integration testing in [`Carts.Api.Tests`](./Carts/Carts.Api.Tests)
  - [API integration tests](./Carts/Carts.Api.Tests/Carts/InitializingCart/InitializeCartTests.cs)

## Other
- [EventTypeMapper](./Core/Core/Events/EventTypeMapper.cs) class to allow both convention-based mapping (by the .NET type name) and custom to handle event versioning,
- [StreamNameMapper](./Core/Core/Events/StreamNameMapper.cs) class for convention-based id (and optional tenant) mapping based on the stream type and module,
- IoC [registration helpers for EventStoreDB configuration](./Core/Core.EventStoreDB/Config.cs),

Besides that:
- Added README.md with repository description
- Added CONTRIBUTING.md with contribution guidelines
- set MIT Licence
- Github Action with build, test and acceptance tests for the sample and [test report](https://github.com/EventStore/samples/runs/3077454539?check_suite_focus=true)

### CODE REVIEW

**Review by ylorph — CHANGES_REQUESTED** [review]

File: README.md [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+Documentation for [EventStoreDB](https://www.eventstore.com/) can be found at: https://developers.eventstore.com/.
+
+We invite you to join our community discussion space at [Discuss](https://discuss.eventstore.com/).
+
+## Contributing
+
+Feel free to [create a GitHub issue](https://github.com/EventStore/samples/issues/new) if you have any questions or request for more explanation or samples.
+
+We're open for any contribution! If you noticed some inconsistency, missing piece, or you'd like to extend existing samples - we'll be happy to [get your Pull Request](https://github.com/EventStore/samples/compare).
```
[the remark this answers is not in the document]
Comment (oskardudycz) [reply README.md:15]: Yes, I wrote about that in the Contribution guidelines.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+2. Once you get approval from one of the maintainers, you can start to work on your code change.
+3. After your changes are ready, make sure that you covered your case with automated tests and verify that you have limited the number of breaking changes to a bare minimum.
+4. Make sure that your code is compiling and all automated tests are passing.
+
+## After you have sent Pull Request
+
+1. Make sure that you applied or answered all the feedback from the maintainers.
+2. We're trying to be as much responsive as we can, but if we didn't respond to you, feel free to ping us by tagging us in the Issue or Pull Request.
+3. Pull request will be merged when you get approvals from at least 2 of the maintainers (and no rejection from others). Pull request will be tagged with the target Marten version in which it will be released. We also label the Pull Requests with information about the type of change.
```
[the remark this answers is not in the document]
Comment (oskardudycz) [reply CONTRIBUTING.md]: Too much of copy/paste ;) Fixed in https://github.com/EventStore/samples/pull/1/commits/21ebc2d614ccbb3fc6b7581d722a1e2ac9f7e1c3.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+1. Fork the repository.
+2. Create a feature branch from the `main` (or release) branch.
+3. It's recommended using rebase strategy for feature branches (see more in [Git documentation](https://git-scm.com/book/en/v2/Git-Branching-Rebasing)). Having that, we highly recommend using clear commit messages. Commits should also represent the unit of change.
+4. Before sending PR to ensure that you rebased the latest source branch from the main repository.
+5. When you're ready to create the [Pull Request on GitHub](https://github.com/EventStore/samples/compare).
+
+## Code style
+
+Coding rules are set up in the project files with `.editorconfig` or `.prettierrc.json`. Those files are supported by all popular IDE (e.g. Microsoft Visual Studio, Rider, Visual Studio Code). Unless you disabled it manually, it should be automatically applied after opening the solution. We also recommend turning automatic formatting on saving to have all the rules applied.
```
[the remark this answers is not in the document]
Comment (oskardudycz) [reply CONTRIBUTING.md]: Reworded that a bit in https://github.com/EventStore/samples/pull/1/commits/21ebc2d614ccbb3fc6b7581d722a1e2ac9f7e1c3. @ylorph could you have a look again?

File: README.md [hunk 4]
```diff
@@ -0,0 +1,40 @@
+# EventStoreDB samples
+
+[EventStoreDB](https://www.eventstore.com/) is an industrial-strength database technology used as the central data store for event-sourced systems. It is available open-source to run locally on most platforms or as SaaS through [Event Store Cloud](https://www.eventstore.com/event-store-cloud).
+
+This repository provides real-world, practical samples both for the [EventStoreDB](https://www.eventstore.com/) and clients SDK.
```
[the remark this answers is not in the document]
Comment (oskardudycz) [reply README.md]: Reworded in https://github.com/EventStore/samples/pull/1/commits/21ebc2d614ccbb3fc6b7581d722a1e2ac9f7e1c3.

**Comment by oskardudycz** [comment]
@ylorph, I think that I applied or answered or your comments. Thanks for checking that :+1: Good feedback.

File: README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (ylorph) [reply README.md:11]: no you're right

**Review by ylorph — APPROVED** [review]

**Review by mat-mcloughlin — APPROVED** [review]

**Pull request merged**